File: OmniPlay-win/src/OmniPlay.Infrastructure/Library/LuckyStunClient.cs
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using OmniPlay.Core.Interfaces;
using OmniPlay.Core.Settings;

namespace OmniPlay.Infrastructure.Library;

public sealed class LuckyStunClient : ILuckyStunClient
{
    private static readonly string[] LoginPaths =
    [
        "/api/login",
        "/api/auth/login",
        "/api/user/login",
        "/api/auth/signin",
        "/login"
    ];

    private static readonly string[] RulePaths =
    [
        "/api/stun",
        "/api/stun/rules",
        "/api/stun/list",
        "/api/penetration",
        "/api/penetration/rules",
        "/api/nat/stun",
        "/api/config/stun",
        "/api/rules",
        "/api/getstunlist"
    ];

    private static readonly string[] IdKeys = ["id", "ruleId", "uuid", "key", "index"];
    private static readonly string[] NameKeys = ["name", "title", "ruleName", "remark", "description"];
    private static readonly string[] AddressKeys =
    [
        "address",
        "url",
        "remoteAddress",
        "publicAddress",
        "externalAddress",
        "penetrationAddress",
        "forwardAddress",
        "stunAddress",
        "remoteUrl",
        "domain",
        "host"
    ];

    private readonly HttpClient httpClient;

    public LuckyStunClient(HttpClient httpClient)
    {
        this.httpClient = httpClient;
    }

    public async Task<LuckyStunSnapshot> FetchAsync(
        LuckyStunSettings settings,
        CancellationToken cancellationToken = default)
    {
        var baseUrl = NormalizeManagementUrl(settings.ManagementUrl);
        if (string.IsNullOrWhiteSpace(baseUrl))
        {
            throw new InvalidOperationException("请填写 Lucky STUN 管理地址。");
        }

        var cookies = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        string? token = null;
        if (!string.IsNullOrWhiteSpace(settings.Username) || !string.IsNullOrWhiteSpace(settings.Password))
        {
            (token, cookies) = await TryLoginAsync(
                baseUrl,
                settings.Username.Trim(),
                settings.Password,
                cookies,
                cancellationToken);
        }

        var lastFailure = "Lucky STUN 没有返回规则列表。";
        foreach (var path in RulePaths)
        {
            try
            {
                var response = await SendAsync(
                    HttpMethod.Get,
                    CombineUrl(baseUrl, path),
                    body: null,
                    token,
                    cookies,
                    cancellationToken);
                if ((int)response.StatusCode is < 200 or >= 300)
                {
                    lastFailure = $"HTTP {(int)response.StatusCode}";
                    continue;
                }

                var rules = ParseRules(response.Body);
                if (rules.Count == 0)
                {
                    lastFailure = "响应中没有识别到穿透规则。";
                    continue;
                }

                var selected = SelectRule(rules, settings);
                return new LuckyStunSnapshot(
                    rules,
                    selected,
                    selected is null
                        ? "已登录，但没有可用的穿透地址。"
                        : $"已读取 {rules.Count} 条规则，当前地址：{selected.Address}");
            }
            catch (HttpRequestException ex)
            {
                lastFailure = ex.Message;
            }
            catch (TaskCanceledException) when (!cancellationToken.IsCancellationRequested)
            {
                lastFailure = "Lucky STUN 请求超时。";
            }
            catch (JsonException ex)
            {
                lastFailure = $"规则响应解析失败：{ex.Message}";
            }
        }

        throw new InvalidOperationException($"无法读取 Lucky STUN 规则：{lastFailure}");
    }

    public static string NormalizeAddress(string value)
    {
        var trimmed = value.Trim();
        if (trimmed.StartsWith("//", StringComparison.Ordinal))
        {
            return $"http:{trimmed}";
        }

        if (trimmed.StartsWith("http://", StringComparison.OrdinalIgnoreCase) ||
            trimmed.StartsWith("https://", StringComparison.OrdinalIgnoreCase))
        {
            return trimmed.TrimEnd('/');
        }

        return trimmed.Contains(':', StringComparison.Ordinal) || trimmed.Contains('.', StringComparison.Ordinal)
            ? $"http://{trimmed.TrimEnd('/') }"
            : trimmed;
    }

    private async Task<(string? Token, HashSet<string> Cookies)> TryLoginAsync(
        string baseUrl,
        string username,
        string password,
        HashSet<string> cookies,
        CancellationToken cancellationToken)
    {
        var body = JsonSerializer.Serialize(new { username, password });
        var lastStatus = HttpStatusCode.BadRequest;
        foreach (var path in LoginPaths)
        {
            try
            {
                var response = await SendAsync(
                    HttpMethod.Post,
                    CombineUrl(baseUrl, path),
                    body,
                    token: null,
                    cookies,
                    cancellationToken);
                lastStatus = response.StatusCode;
                if ((int)response.StatusCode is < 200 or >= 300)
                {
                    continue;
                }

                var token = TryFindString(response.Body, ["token", "accessToken", "access_token", "jwt", "authorization"]);
                if (!string.IsNullOrWhiteSpace(token))
                {
                    token = token.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase)
                        ? token
                        : $"Bearer {token}";
                }

                return (token, cookies);
            }
            catch (HttpRequestException)
            {
            }
            catch (TaskCanceledException) when (!cancellationToken.IsCancellationRequested)
            {
            }
        }

        if (lastStatus is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
        {
            throw new UnauthorizedAccessException("Lucky STUN 登录失败，请检查账号和密码。");
        }

        return (null, cookies);
    }

    private async Task<LuckyHttpResponse> SendAsync(
        HttpMethod method,
        string url,
        string? body,
        string? token,
        HashSet<string> cookies,
        CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(method, url);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        request.Headers.UserAgent.ParseAdd("OmniPlay-LuckySTUN");
        if (!string.IsNullOrWhiteSpace(token))
        {
            request.Headers.Authorization = AuthenticationHeaderValue.Parse(token);
        }

        if (cookies.Count > 0)
        {
            request.Headers.TryAddWithoutValidation("Cookie", string.Join("; ", cookies));
        }

        if (body is not null)
        {
            request.Content = new StringContent(body, Encoding.UTF8, "application/json");
        }

        using var response = await httpClient.SendAsync(request, HttpCompletionOption.ResponseContentRead, cancellationToken);
        foreach (var cookie in response.Headers.TryGetValues("Set-Cookie", out var values) ? values : [])
        {
            var value = cookie.Split(';', 2)[0].Trim();
            if (!string.IsNullOrWhiteSpace(value))
            {
                cookies.RemoveWhere(existing => existing.StartsWith(value.Split('=', 2)[0] + "=", StringComparison.OrdinalIgnoreCase));
                cookies.Add(value);
            }
        }

        return new LuckyHttpResponse(response.StatusCode, await response.Content.ReadAsStringAsync(cancellationToken));
    }

    private static IReadOnlyList<LuckyStunRule> ParseRules(string body)
    {
        using var document = JsonDocument.Parse(body);
        var rules = new List<LuckyStunRule>();
        CollectRules(document.RootElement, rules, "rule");
        return rules
            .GroupBy(static rule => $"{rule.Id}\n{rule.Name}\n{rule.Address}", StringComparer.OrdinalIgnoreCase)
            .Select(static group => group.First())
            .ToArray();
    }

    private static void CollectRules(JsonElement element, ICollection<LuckyStunRule> rules, string fallbackName)
    {
        if (element.ValueKind == JsonValueKind.Array)
        {
            foreach (var child in element.EnumerateArray())
            {
                CollectRules(child, rules, fallbackName);
            }

            return;
        }

        if (element.ValueKind != JsonValueKind.Object)
        {
            return;
        }

        var id = FindString(element, IdKeys) ?? string.Empty;
        var name = FindString(element, NameKeys) ?? fallbackName;
        var address = FindAddress(element);
        if (!string.IsNullOrWhiteSpace(address) && LooksLikeAddress(address))
        {
            rules.Add(new LuckyStunRule(
                string.IsNullOrWhiteSpace(id) ? $"{name}:{address}" : id,
                string.IsNullOrWhiteSpace(name) ? "Lucky STUN 规则" : name,
                NormalizeAddress(address)));
        }

        foreach (var property in element.EnumerateObject())
        {
            if (property.Value.ValueKind is JsonValueKind.Object or JsonValueKind.Array)
            {
                CollectRules(property.Value, rules, name);
            }
        }
    }

    private static string? FindAddress(JsonElement element)
    {
        foreach (var key in AddressKeys)
        {
            var value = FindString(element, [key]);
            if (!string.IsNullOrWhiteSpace(value))
            {
                return value;
            }
        }

        var host = FindString(element, ["remoteHost", "publicHost", "externalHost"]);
        var port = FindString(element, ["remotePort", "publicPort", "externalPort", "port"]);
        return string.IsNullOrWhiteSpace(host) || string.IsNullOrWhiteSpace(port)
            ? null
            : $"{host}:{port}";
    }

    private static string? FindString(JsonElement element, IReadOnlyList<string> keys)
    {
        foreach (var property in element.EnumerateObject())
        {
            if (!keys.Any(key => string.Equals(property.Name, key, StringComparison.OrdinalIgnoreCase)))
            {
                continue;
            }

            return property.Value.ValueKind switch
            {
                JsonValueKind.String => property.Value.GetString(),
                JsonValueKind.Number => property.Value.GetRawText(),
                JsonValueKind.True => "true",
                JsonValueKind.False => "false",
                _ => null
            };
        }

        return null;
    }

    private static bool LooksLikeAddress(string value)
    {
        var trimmed = value.Trim();
        return (trimmed.Contains('.', StringComparison.Ordinal) || trimmed.Contains(':', StringComparison.Ordinal)) &&
               !trimmed.Contains("/api/", StringComparison.OrdinalIgnoreCase);
    }

    private static string? TryFindString(string body, IReadOnlyList<string> keys)
    {
        try
        {
            using var document = JsonDocument.Parse(body);
            return FindStringRecursive(document.RootElement, keys);
        }
        catch (JsonException)
        {
            return null;
        }
    }

    private static string? FindStringRecursive(JsonElement element, IReadOnlyList<string> keys)
    {
        if (element.ValueKind == JsonValueKind.Object)
        {
            var direct = FindString(element, keys);
            if (!string.IsNullOrWhiteSpace(direct))
            {
                return direct;
            }

            foreach (var property in element.EnumerateObject())
            {
                var nested = FindStringRecursive(property.Value, keys);
                if (!string.IsNullOrWhiteSpace(nested))
                {
                    return nested;
                }
            }
        }
        else if (element.ValueKind == JsonValueKind.Array)
        {
            foreach (var child in element.EnumerateArray())
            {
                var nested = FindStringRecursive(child, keys);
                if (!string.IsNullOrWhiteSpace(nested))
                {
                    return nested;
                }
            }
        }

        return null;
    }

    private static LuckyStunRule? SelectRule(IReadOnlyList<LuckyStunRule> rules, LuckyStunSettings settings)
    {
        if (!string.IsNullOrWhiteSpace(settings.SelectedRuleId))
        {
            var byId = rules.FirstOrDefault(rule => string.Equals(rule.Id, settings.SelectedRuleId.Trim(), StringComparison.OrdinalIgnoreCase));
            if (byId is not null)
            {
                return byId;
            }
        }

        if (!string.IsNullOrWhiteSpace(settings.SelectedRuleName))
        {
            var byName = rules.FirstOrDefault(rule => string.Equals(rule.Name, settings.SelectedRuleName.Trim(), StringComparison.OrdinalIgnoreCase));
            if (byName is not null)
            {
                return byName;
            }
        }

        return rules.FirstOrDefault();
    }

    private static string NormalizeManagementUrl(string value)
    {
        var trimmed = value.Trim();
        if (trimmed.Length == 0)
        {
            return string.Empty;
        }

        if (!trimmed.Contains("://", StringComparison.Ordinal))
        {
            trimmed = $"http://{trimmed}";
        }

        if (!Uri.TryCreate(trimmed, UriKind.Absolute, out var uri) ||
            uri.Scheme is not ("http" or "https") ||
            string.IsNullOrWhiteSpace(uri.Host))
        {
            return string.Empty;
        }

        var path = uri.AbsolutePath.TrimEnd('/');
        if (path.EndsWith("/api", StringComparison.OrdinalIgnoreCase))
        {
            path = path[..^4].TrimEnd('/');
        }

        var builder = new UriBuilder(uri) { Path = path, Query = string.Empty, Fragment = string.Empty };
        return builder.Uri.AbsoluteUri.TrimEnd('/');
    }

    private static string CombineUrl(string baseUrl, string path)
    {
        return $"{baseUrl.TrimEnd('/')}/{path.TrimStart('/')}";
    }

    private sealed record LuckyHttpResponse(HttpStatusCode StatusCode, string Body);
}

File: OmniPlay-安卓TV/app/src/main/java/com/omniplay/tv/data/LuckyStunClient.java
package com.omniplay.tv.data;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public final class LuckyStunClient {
    private static final String[] LOGIN_PATHS = {
            "/api/login", "/api/auth/login", "/api/user/login", "/api/auth/signin", "/login"
    };
    private static final String[] RULE_PATHS = {
            "/api/stun", "/api/stun/rules", "/api/stun/list", "/api/penetration",
            "/api/penetration/rules", "/api/nat/stun", "/api/config/stun", "/api/rules", "/api/getstunlist"
    };

    private LuckyStunClient() {
    }

    public static Result fetch(String managementUrl, String username, String password, String selectedRuleId, String selectedRuleName)
            throws IOException, JSONException {
        String baseUrl = normalizeManagementUrl(managementUrl);
        if (baseUrl.isEmpty()) {
            throw new IOException("Lucky STUN 管理地址无效。");
        }

        ArrayList<String> cookies = new ArrayList<>();
        String token = "";
        if (!trim(username).isEmpty() || password != null && !password.isEmpty()) {
            LoginResult login = login(baseUrl, trim(username), password == null ? "" : password, cookies);
            token = login.token;
            cookies = login.cookies;
        }

        String lastError = "没有识别到规则列表。";
        for (String path : RULE_PATHS) {
            try {
                Response response = request("GET", baseUrl + path, null, token, cookies);
                if (response.status < 200 || response.status >= 300) {
                    lastError = "HTTP " + response.status;
                    continue;
                }
                List<Rule> rules = parseRules(response.body);
                if (!rules.isEmpty()) {
                    Rule selected = selectRule(rules, selectedRuleId, selectedRuleName);
                    return new Result(rules, selected, "已读取 " + rules.size() + " 条规则。");
                }
                lastError = "响应中没有识别到穿透规则。";
            } catch (JSONException error) {
                lastError = "规则响应解析失败：" + error.getMessage();
            }
        }
        throw new IOException("无法读取 Lucky STUN 规则：" + lastError);
    }

    public static String normalizeAddress(String value) {
        String trimmed = trim(value);
        if (trimmed.isEmpty()) {
            return "";
        }
        if (!trimmed.startsWith("http://") && !trimmed.startsWith("https://")) {
            trimmed = "http://" + trimmed;
        }
        try {
            URL url = new URL(trimmed);
            if (url.getHost() == null || url.getHost().isEmpty()) {
                return "";
            }
        } catch (Exception ignored) {
            return "";
        }
        while (trimmed.endsWith("/")) {
            trimmed = trimmed.substring(0, trimmed.length() - 1);
        }
        return trimmed;
    }

    private static LoginResult login(String baseUrl, String username, String password, ArrayList<String> cookies) throws IOException, JSONException {
        JSONObject body = new JSONObject();
        body.put("username", username);
        body.put("password", password);
        int lastStatus = 400;
        for (String path : LOGIN_PATHS) {
            try {
                Response response = request("POST", baseUrl + path, body.toString(), "", cookies);
                cookies = mergeCookies(cookies, response.cookies);
                lastStatus = response.status;
                if (response.status < 200 || response.status >= 300) {
                    continue;
                }
                String token = findStringRecursive(response.body, new String[]{"token", "accessToken", "access_token", "jwt", "authorization"});
                if (token != null && !token.toLowerCase(Locale.ROOT).startsWith("bearer ")) {
                    token = "Bearer " + token;
                }
                return new LoginResult(token == null ? "" : token, cookies);
            } catch (IOException ignored) {
            }
        }
        if (lastStatus == 401 || lastStatus == 403) {
            throw new IOException("Lucky STUN 登录失败，请检查账号和密码。");
        }
        return new LoginResult("", cookies);
    }

    private static Response request(String method, String url, String body, String token, List<String> cookies) throws IOException {
        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
        connection.setConnectTimeout(6000);
        connection.setReadTimeout(12000);
        connection.setRequestMethod(method);
        connection.setRequestProperty("Accept", "application/json");
        connection.setRequestProperty("User-Agent", "OmniPlay-Android-LuckySTUN");
        if (token != null && !token.isEmpty()) {
            connection.setRequestProperty("Authorization", token);
        }
        if (cookies != null && !cookies.isEmpty()) {
            connection.setRequestProperty("Cookie", join(cookies, "; "));
        }
        if (body != null) {
            byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
            connection.setDoOutput(true);
            connection.setRequestProperty("Content-Type", "application/json; charset=utf-8");
            connection.setFixedLengthStreamingMode(bytes.length);
            try (OutputStream output = connection.getOutputStream()) {
                output.write(bytes);
            }
        }
        int status = connection.getResponseCode();
        InputStream stream = status >= 400 ? connection.getErrorStream() : connection.getInputStream();
        String text = readText(stream);
        ArrayList<String> responseCookies = new ArrayList<>();
        for (var entry : connection.getHeaderFields().entrySet()) {
            if (entry.getKey() != null && "Set-Cookie".equalsIgnoreCase(entry.getKey()) && entry.getValue() != null) {
                for (String header : entry.getValue()) {
                    if (header != null && !header.isEmpty()) {
                        responseCookies.add(header.split(";", 2)[0]);
                    }
                }
            }
        }
        connection.disconnect();
        return new Response(status, text, responseCookies);
    }

    private static List<Rule> parseRules(String body) throws JSONException {
        Object root = body.trim().startsWith("[") ? new JSONArray(body) : new JSONObject(body);
        ArrayList<Rule> rules = new ArrayList<>();
        collectRules(root, "Lucky STUN 规则", rules);
        Set<String> seen = new HashSet<>();
        ArrayList<Rule> unique = new ArrayList<>();
        for (Rule rule : rules) {
            String key = (rule.id + "|" + rule.name + "|" + rule.address).toLowerCase(Locale.ROOT);
            if (seen.add(key)) {
                unique.add(rule);
            }
        }
        return unique;
    }

    private static void collectRules(Object value, String fallbackName, List<Rule> rules) throws JSONException {
        if (value instanceof JSONArray) {
            JSONArray array = (JSONArray) value;
            for (int i = 0; i < array.length(); i++) {
                collectRules(array.get(i), fallbackName, rules);
            }
            return;
        }
        if (!(value instanceof JSONObject)) {
            return;
        }
        JSONObject object = (JSONObject) value;
        String id = findString(object, new String[]{"id", "ruleId", "uuid", "key", "index"});
        String name = findString(object, new String[]{"name", "title", "ruleName", "remark", "description"});
        if (name == null || name.isEmpty()) {
            name = fallbackName;
        }
        String address = findAddress(object);
        if (address != null && (address.contains(".") || address.contains(":"))) {
            String normalized = normalizeAddress(address);
            if (!normalized.isEmpty()) {
                rules.add(new Rule(id == null || id.isEmpty() ? name + ":" + normalized : id, name, normalized));
            }
        }
        Iterator<String> keys = object.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            Object child = object.opt(key);
            if (child instanceof JSONObject || child instanceof JSONArray) {
                collectRules(child, name, rules);
            }
        }
    }

    private static String findAddress(JSONObject object) {
        String value = findString(object, new String[]{"address", "url", "remoteAddress", "publicAddress", "externalAddress", "penetrationAddress", "forwardAddress", "stunAddress", "remoteUrl", "domain", "host"});
        if (value != null && !value.toLowerCase(Locale.ROOT).contains("/api/")) {
            return value;
        }
        String host = findString(object, new String[]{"remoteHost", "publicHost", "externalHost"});
        String port = findString(object, new String[]{"remotePort", "publicPort", "externalPort", "port"});
        return host == null || host.isEmpty() || port == null || port.isEmpty() ? null : host + ":" + port;
    }

    private static Rule selectRule(List<Rule> rules, String id, String name) {
        String selectedId = trim(id);
        String selectedName = trim(name);
        for (Rule rule : rules) {
            if (!selectedId.isEmpty() && rule.id.equalsIgnoreCase(selectedId)) {
                return rule;
            }
        }
        for (Rule rule : rules) {
            if (!selectedName.isEmpty() && rule.name.equalsIgnoreCase(selectedName)) {
                return rule;
            }
        }
        return rules.isEmpty() ? null : rules.get(0);
    }

    private static String findString(JSONObject object, String[] keys) {
        Iterator<String> iterator = object.keys();
        while (iterator.hasNext()) {
            String name = iterator.next();
            for (String key : keys) {
                if (!name.equalsIgnoreCase(key)) {
                    continue;
                }
                Object value = object.opt(name);
                if (value instanceof String || value instanceof Number || value instanceof Boolean) {
                    return String.valueOf(value);
                }
            }
        }
        return null;
    }

    private static String findStringRecursive(String body, String[] keys) {
        try {
            Object root = body.trim().startsWith("[") ? new JSONArray(body) : new JSONObject(body);
            return findStringRecursive(root, keys);
        } catch (JSONException ignored) {
            return null;
        }
    }

    private static String findStringRecursive(Object value, String[] keys) throws JSONException {
        if (value instanceof JSONObject) {
            JSONObject object = (JSONObject) value;
            String direct = findString(object, keys);
            if (direct != null && !direct.isEmpty()) {
                return direct;
            }
            Iterator<String> iterator = object.keys();
            while (iterator.hasNext()) {
                String nested = findStringRecursive(object.opt(iterator.next()), keys);
                if (nested != null && !nested.isEmpty()) {
                    return nested;
                }
            }
        } else if (value instanceof JSONArray) {
            JSONArray array = (JSONArray) value;
            for (int i = 0; i < array.length(); i++) {
                String nested = findStringRecursive(array.get(i), keys);
                if (nested != null && !nested.isEmpty()) {
                    return nested;
                }
            }
        }
        return null;
    }

    private static ArrayList<String> mergeCookies(List<String> old, List<String> fresh) {
        ArrayList<String> result = new ArrayList<>(old == null ? java.util.Collections.emptyList() : old);
        for (String cookie : fresh) {
            String key = cookie.split("=", 2)[0];
            result.removeIf(item -> item.startsWith(key + "="));
            result.add(cookie);
        }
        return result;
    }

    private static String normalizeManagementUrl(String value) {
        String trimmed = trim(value);
        if (trimmed.isEmpty()) {
            return "";
        }
        if (!trimmed.contains("://")) {
            trimmed = "http://" + trimmed;
        }
        try {
            URL url = new URL(trimmed);
            if (!(url.getProtocol().equalsIgnoreCase("http") || url.getProtocol().equalsIgnoreCase("https")) || url.getHost().isEmpty()) {
                return "";
            }
        } catch (Exception ignored) {
            return "";
        }
        while (trimmed.endsWith("/")) {
            trimmed = trimmed.substring(0, trimmed.length() - 1);
        }
        if (trimmed.toLowerCase(Locale.ROOT).endsWith("/api")) {
            trimmed = trimmed.substring(0, trimmed.length() - 4);
            while (trimmed.endsWith("/")) {
                trimmed = trimmed.substring(0, trimmed.length() - 1);
            }
        }
        return trimmed;
    }

    private static String join(List<String> values, String separator) {
        StringBuilder builder = new StringBuilder();
        for (String value : values) {
            if (builder.length() > 0) {
                builder.append(separator);
            }
            builder.append(value);
        }
        return builder.toString();
    }

    private static String trim(String value) {
        return value == null ? "" : value.trim();
    }

    private static String readText(InputStream stream) throws IOException {
        if (stream == null) {
            return "";
        }
        StringBuilder builder = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) {
                builder.append(line);
            }
        }
        return builder.toString();
    }

    public static final class Rule {
        public final String id;
        public final String name;
        public final String address;

        Rule(String id, String name, String address) {
            this.id = id;
            this.name = name;
            this.address = address;
        }
    }

    public static final class Result {
        public final List<Rule> rules;
        public final Rule selectedRule;
        public final String message;

        Result(List<Rule> rules, Rule selectedRule, String message) {
            this.rules = rules;
            this.selectedRule = selectedRule;
            this.message = message;
        }
    }

    private static final class LoginResult {
        final String token;
        final ArrayList<String> cookies;

        LoginResult(String token, ArrayList<String> cookies) {
            this.token = token;
            this.cookies = cookies;
        }
    }

    private static final class Response {
        final int status;
        final String body;
        final ArrayList<String> cookies;

        Response(int status, String body, ArrayList<String> cookies) {
            this.status = status;
            this.body = body;
            this.cookies = cookies;
        }
    }
}
